File: docker-environment/js-docker-demo/logger.js
let testCases = [];
let totalCorrect = 0;
let totalTestcases = 0;

const logTestCase = (index, stats, inputs, expectedOutput) => {
  totalTestcases++;
  if (stats.passed) totalCorrect++;

  const testCaseResult = {
    test_case: index,
    passed: stats.passed,
    inputs: JSON.stringify(inputs),
    output: stats.result === undefined ? "undefined" : JSON.stringify(stats.result),
    expected: JSON.stringify(expectedOutput),
    run_time: `${stats.runTime.toFixed(3)} ms`,
    memory: `${(stats.memoryUsed / 1024).toFixed(3)} KB`
  };

  testCases.push(testCaseResult);
};

const logSummary = (stats) => {
  const summary = {
    run_result: testCases,
    total_correct: totalCorrect,
    total_testcases: totalTestcases,
    total_run_time: `${stats.totalRunTime.toFixed(3)} ms`,
    total_run_memory: `${(stats.totalMemoryUsed / 1024).toFixed(3)} MB`,
    all_passed: stats.allTestsPassed
  };

  console.log(JSON.stringify(summary, (key, value) => 
    value === undefined ? "undefined" : value
  ));
};

module.exports = { logTestCase, logSummary };

File: docker-environment/python-docker-demo/logger.py
import json

test_cases = []
total_correct = 0
total_testcases = 0


def log_test_case(index, stats, inputs, expected_output):
    global total_correct, total_testcases
    total_testcases += 1
    if stats["passed"]:
        total_correct += 1

    test_case_result = {
        "test_case": index,
        "passed": stats["passed"],
        "inputs": json.dumps(inputs),
        "output": (
            "undefined" if stats["result"] is None else json.dumps(stats["result"])
        ),
        "expected": json.dumps(expected_output),
        "run_time": f"{stats['run_time']:.3f} ms",
        "memory": f"{stats['memory_used'] / 1024:.3f} KB",
    }

    test_cases.append(test_case_result)


def log_summary(stats):
    summary = {
        "run_result": test_cases,
        "total_correct": total_correct,
        "total_testcases": total_testcases,
        "total_run_time": f"{stats['total_run_time']:.3f} ms",
        "total_run_memory": f"{stats['total_memory_used'] / 1024:.3f} MB",
        "all_passed": stats["all_tests_passed"],
    }

    print(json.dumps(summary, default=lambda x: "undefined" if x is None else x))
